specification/dartLangSpec.tex

```diff
@@ -6281,66 +6293,120 @@ \subsection{Constants}
-\begin{enumerate}
-\item It is constant.
-We can evaluate it at compile time.
-\item It seems more useful to give the error stemming from the evaluation explicitly.
-\end{enumerate}
+First, it is constant so we \emph{can} evaluate it at compile time.
+Second, it seems more useful to give
+the error stemming from the evaluation explicitly.%
 }
 
 \rationale{
-One might reasonably ask why $e_1$\,?\,$e_1$\,:\,$e_3$ and $e_1$\,??\,$e_2$ have constant forms.
-For example, if $e_1$ is known statically, why do we need to test it?
-The answer is that there are contexts where $e_1$ is a variable.
-In particular, constant constructor initializers such as
+One might reasonably ask why
+$e_1$\,?\,\,$e_2$\,:\,$e_3$ and $e_1$\,??\,\,$e_2$
+have constant forms.
+If $e_1$ is known statically, why do we need to test it?
+The answer is that there are contexts where $e_1$ is a variable,
+e.g., in constant constructor initializers such as
+\code{\CONST{} C(foo):\ \THIS.foo = foo ??\ someDefaultValue;}
+}
+
+\commentary{
+The difference between
+a potentially constant expression and a constant expression
+deserves some explanation.
+The key issue is how one treats the formal parameters of a constructor.
+
+If a constant constructor is invoked from a constant object expression,
+the actual arguments will be required to be constant expressions.
+Therefore, if we were assured that
+constant constructors were always invoked from constant object expressions,
+we could assume that the formal parameters of a constructor were
+compile-time constants.
 
-\code{\CONST{} C(foo): \THIS.foo = foo ?? someDefaultValue;}
+However, constant constructors can also be invoked from
+ordinary instance creation expressions (\ref{new}),
+and so the above assumption is not generally valid.
+
+Nevertheless, the use of the formal parameters of a constant constructor
+is of considerable utility.
+The concept of potentially constant expressions is introduced to facilitate
+limited use of such formal parameters.
+Specifically, we allow the usage of
+the formal parameters of a constant constructor
+for expressions that involve built-in operators,
+but not for constant objects, lists and maps.
+For instance:
 }
 
-\LMHash{}%
-It is a compile-time error if the value of a constant expression depends on itself.
+\begin{dartCode}
+\CLASS{} C \{
+  \FINAL{} x, y, z;
+  \CONST{} C(p, q): x = q, y = p + 100, z = p + q;
+\}
+\end{dartCode}
 
 \commentary{
-As an example, consider:
+The assignment to \code{x} is allowed under the assumption that \code{q} is constant (even though \code{q} is not, in general a compile-time constant).
+The assignment to \code{y} is similar, but raises additional questions.
+In this case, the superexpression of \code{p} is \code{p + 100}, and it requires that \code{p} be a numeric constant expression for the entire expression to be considered constant.
+The wording of the specification allows us to assume that \code{p} evaluates to an integer.
```

Comment:
In a constant setting, but not in a non-constant invocation.

Reply:
Added text to indicate this.